incrementals-publisher/index.js

```diff
@@ -0,0 +1,142 @@
+/*
+ * This Azure Function is responsible for processing information related to an
+ * incrementals release and bouncing the artifacts into Artifactory
+ */
+
+const fs    = require('fs');
+const fetch = require('node-fetch');
+const os    = require('os');
+const path  = require('path');
+const util  = require('util');
+
+const github      = require('./lib/github');
+const pipeline    = require('./lib/pipeline');
+const permissions = require('./lib/permissions');
+
+const JENKINS_HOST     = process.env.JENKINS_HOST || 'https://ci.jenkins.io';
+const INCREMENTAL_URL  = process.env.INCREMENTAL_URL || 'https://repo.jenkins-ci.org/incrementals/'
+const ARTIFACTORY_KEY  = process.env.ARTIFACTORY_KEY || 'invalid-key';
+
+const TEMP_ARCHIVE_DIR = path.join(os.tmpdir(), 'incrementals-');
+const mktemp           = util.promisify(fs.mkdtemp);
+
+/*
+ * Small helper function to make failing a request more concise
+ */
+const failRequest = (context, body) => {
+  context.res = {
+    status: 500,
+    body: body || 'Unknown error'
+  };
+  context.done();
+};
+
+
+module.exports = async (context, data) => {
+  const buildUrl = data.body.build_url;
+  /* If we haven't received any valid data, just bail early
+   */
+  if ((!buildUrl) || (!buildUrl.match(JENKINS_HOST))) {
+    context.res = {
+      status: 400,
+      body: 'The incrementals-publisher invocation was poorly formed and missing attributes'
+    };
+    context.done();
+    return;
+  }
+  // Starting some async operations early which we will need later
+  let tmpDir = mktemp(TEMP_ARCHIVE_DIR);
+  let perms = permissions.fetch();
+
+  /*
+   * The first step is to take the buildUrl and fetch some metadata about this
+   * specific Pipeline Run
+   */
+  let metdata_url = pipeline.getApiUrl(buildUrl);
+  if (process.env.METADATA_URL) {
+    metadataUrl = process.env.METADATA_URL;
+    context.log.info('Using an override for the metadata URL:', metadataUrl);
+  }
+  let metadata = await fetch(metadataUrl);
+
+  if (metadata.status != 200) {
+    context.log.error('Failed to fetch Pipeline metadata', metadata.body);
+  }
+  metadata = await metadata.json();
+
+  if (!metadata) {
+    context.log.error('I was unable to parse any JSON metadata', metadata);
+    return failRequest(context);
+  }
+  metadata = pipeline.processMetadata(metadata);
```

Comment:
Wait, you are switching from a Jenkins JSON to a hash of an unrelated type? That deserves a fresh variable no?